proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,370 @@
+- `-enable-stable-abi`
+- just `-stable-abi`
+
+In practice, the precise name of this flag won't be very important, because (1) the number of people outside of Apple making libraries with binary stability concerns won't be very high, and (2) most people will be building those libraries through Xcode or SwiftPM, which will have its own name for this mode. But naming the *feature* is still important.
+
+### "No reordering" vs. "No renaming"
+
+As written, this proposal disallows *reordering* field of a fixed-contents struct. This matches the restrictions of C structs, but is different from everything else in Swift, which freely allows reordering. Why?
+
+Let's step back and remember our goal: a stable, unchanging ABI for a fixed-contents struct. That means that whatever changes are made to the struct, the in-memory representation and the rest of the ABI must be consistent across library versions. We can think of this as a function that takes fields and produces a layout:
+
+```swift
+struct Field {
+  var name: String
+  var type: ValueType
+}
+typealias Offset = Int
+func layoutForFixedContentsStruct(fields: [Field]) -> [Offset: Field]
+```
+
+To make things simple, let's assume all the fields in the struct have the same type, like this PolarPoint:
+
+```swift
+@fixedContents public struct PolarPoint {
+  // Yes, these have different access levels.
+  // You probably wouldn't do that in real life,
+  // but it's needed for this contrived example.
+  public var radius: Double
+  private var angle: Double
+
+  public init(x: Double, y: Double) {
+    self.radius = sqrt(x * x, y * y)
+    self.angle = atan2(y, x)
+  }
+
+  public var x: Double {
+    return radius * cos(angle)
+  }
+  public var y: Double {
+    return radius * sin(angle)
+  }
+}
+```
+
+We have two choices for how to lay out the PolarPoint struct:
+
+- `[0: radius, 8: angle]`
+- `[0: angle, 8: radius]`
+
+*It's important that everyone using the struct agrees on which layout we're using.* Otherwise, an angle could get misinterpreted as a radius when someone outside the module tries to access the `radius` property.
+
+So how do we decide? We have three choices:
+
+1. Sort the fields by declaration order.
+2. Sort the fields by name.
+3. Do something clever with access levels.
+
+Which translates to three possible restrictions:
+
+1. Fields of a fixed-contents struct may not be reordered, even non-public ones.
+2. Fields of a fixed-contents struct may not be renamed, even non-public ones.
+3. Copying the struct requires calling a function because we can't see the private fields.
+
+(3) is a non-starter, since it defeats the performance motivation for `@fixedContents`. That leaves (1) and (2), and we decided it would be *really weird* if renaming the private property `angle` to `theta` changed the layout of the struct. So the restriction on reordering fields was considered the least bad solution. It helps that it matches the behavior of C, Rust, and other languages.
```

Comment:
```suggestion
(3) is a non-starter, since it defeats the performance motivation for `@frozen`. That leaves (1) and (2), and we decided it would be *really weird* if renaming the private property `angle` to `theta` changed the layout of the struct. So the restriction on reordering fields was considered the least bad solution. It helps that it matches the behavior of C, Rust, and other languages.
```